File: src/lib/events.ts
/**
 * Tauri events → react-query invalidation.
 *
 * Events are invalidation notices only (src-tauri/src/events.rs): payloads
 * name cycles and never carry business data, so the database stays the single
 * source of truth (docs/architecture.md, decision D4). Only drag reordering
 * gets optimistic updates; everything else waits for these notices.
 *
 * Query key convention — feature modules MUST use the factories below and
 * never hand-write key arrays; the invalidation logic matches on these shapes:
 *
 *   qk.plannerState()               ["planner-state"]
 *   qk.editorWorkspace(cycleId)     ["editor-workspace", cycleId]
 *   qk.editorWorkspaces(cycleIds)   ["editor-workspaces", cycleIds]   (batch)
 *   qk.previewSummary(cycleId)      ["preview-summary", cycleId]
 *   qk.settings()                   ["settings"]
 *
 * Invalidation matrix (event names mirror the Rust constants):
 *   cycles:changed    {cycle_ids} → planner-state; editor-workspace per id;
 *                     the editor-workspaces batch root — batch responses embed
 *                     per-cycle data under a list key, so there is no
 *                     per-cycle key to target and the whole root is
 *                     invalidated (coarse but correct; refetching is cheap).
 *   tasks:changed     {cycle_ids} → planner-state; editor-workspace per id
 *                     (plus the batch root, as above); preview-summary per id
 *                     — task writes can rewrite proposal snapshots.
 *   proposals:changed {cycle_id}  → preview-summary for that cycle.
 *   ["settings"] is intentionally never touched here: no settings event
 *   exists (set_week_start_day emits nothing), so its mutators invalidate
 *   the settings query themselves.
 */
import type { QueryClient } from "@tanstack/react-query";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

/** Mirrors `events::CycleIdsPayload`. */
export interface CycleIdsPayload {
  cycle_ids: string[];
}

/** Mirrors `events::CycleIdPayload`. */
export interface CycleIdPayload {
  cycle_id: string;
}

export const qk = {
  plannerState: () => ["planner-state"] as const,
  editorWorkspace: (cycleId: string) => ["editor-workspace", cycleId] as const,
  editorWorkspaces: (cycleIds: string[]) => ["editor-workspaces", cycleIds] as const,
  previewSummary: (cycleId: string) => ["preview-summary", cycleId] as const,
  sessions: (dayCycleId: string) => ["sessions", dayCycleId] as const,
  agentConversation: (cycleId: string) => ["agent-conversation", cycleId] as const,
  issueReport: (cycleId: string) => ["issue-report", cycleId] as const,
  settings: () => ["settings"] as const,
  agentActions: (cycleId: string) => ["agent-actions", cycleId] as const,
};

/**
 * Subscribes to the backend change events and wires them to cache
 * invalidation. Returns a cleanup function that unlistens everything; call it
 * once at app teardown.
 */
export async function initEventInvalidation(queryClient: QueryClient): Promise<() => void> {
  const unlisteners: UnlistenFn[] = [];
  try {
    unlisteners.push(await listen("agent:actions_changed", () => invalidateAgentEffects(queryClient)));
    unlisteners.push(
      await listen<CycleIdsPayload>("cycles:changed", (event) => {
        invalidateCycles(queryClient, event.payload.cycle_ids, { taskWrites: false });
      }),
    );
    unlisteners.push(
      await listen<CycleIdsPayload>("tasks:changed", (event) => {
        invalidateCycles(queryClient, event.payload.cycle_ids, { taskWrites: true });
      }),
    );
    unlisteners.push(
      await listen<CycleIdPayload>("proposals:changed", (event) => {
        queryClient.invalidateQueries({ queryKey: qk.previewSummary(event.payload.cycle_id) });
      }),
    );
    unlisteners.push(
      await listen<{ conversation_id: string; cycle_id: string; revision: number }>(
        "agent:conversation_updated",
        (event) => {
          invalidateAgentEffects(queryClient);
          queryClient.invalidateQueries({
            queryKey: qk.agentConversation(event.payload.cycle_id),
          });
        },
      ),
    );
  } catch (err) {
    // Never leak the listeners acquired before the failure.
    for (const unlisten of unlisteners) unlisten();
    throw err;
  }
  return () => {
    for (const unlisten of unlisteners) unlisten();
  };
}

/** A tool turn may stage edits in another cycle, including before an error.
 * Invalidate existing projections after every settled turn; no second state store. */
export function invalidateAgentEffects(queryClient: QueryClient): void {
  for (const root of ["planner-state", "editor-workspace", "editor-workspaces", "preview-summary", "issue-report", "agent-actions", "settings", "app-flag", "ai-settings", "ai-availability", "agent-conversation", "sessions", "calendar", "calendar-range", "schedule-overlaps", "reminders", "repeats", "time-budget", "daily-capacity"]) {
    void queryClient.invalidateQueries({ queryKey: [root] });
  }
}

/** Shared body of the two `{ cycle_ids }` events; see the matrix above. */
function invalidateCycles(
  queryClient: QueryClient,
  cycleIds: string[],
  opts: { taskWrites: boolean },
): void {
  if (cycleIds.length === 0) return; // emitters skip empty sets; stay defensive
  queryClient.invalidateQueries({ queryKey: qk.plannerState() });
  queryClient.invalidateQueries({ queryKey: ["issue-report"] });
  // Length-1 root prefix-matches every batch key built by qk.editorWorkspaces.
  queryClient.invalidateQueries({ queryKey: ["editor-workspaces"] });
  if (opts.taskWrites) queryClient.invalidateQueries({ queryKey: ["editor-workspace"] });
  for (const cycleId of cycleIds) {
    queryClient.invalidateQueries({ queryKey: qk.editorWorkspace(cycleId) });
    // Session mutations (add/start/finish/repeat) emit cycles:changed for the
    // day, so the day's focus-block listing refreshes alongside the columns.
    queryClient.invalidateQueries({ queryKey: qk.sessions(cycleId) });
    if (opts.taskWrites) {
      queryClient.invalidateQueries({ queryKey: qk.previewSummary(cycleId) });
    }
  }
}

File: src-tauri/src/events.rs
//! Typed event emitters.
//!
//! Events are *invalidation notices only*: the payload never carries business
//! data, so the database stays the single source of truth
//! (see `docs/architecture.md`, decision D4).

use serde::Serialize;
use tauri::{Emitter, Runtime};

pub const CYCLES_CHANGED: &str = "cycles:changed";
pub const TASKS_CHANGED: &str = "tasks:changed";
pub const PROPOSALS_CHANGED: &str = "proposals:changed";

#[derive(Debug, Clone, Serialize)]
pub struct CycleIdsPayload {
    pub cycle_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CycleIdPayload {
    pub cycle_id: String,
}

/// Deduplicated set of cycle ids collected during one use case, emitted once.
#[derive(Debug, Default)]
pub struct CycleIdSet(Vec<String>);

impl CycleIdSet {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn push(&mut self, id: impl Into<String>) {
        let id = id.into();
        if !self.0.iter().any(|existing| existing == &id) {
            self.0.push(id);
        }
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    pub fn ids(&self) -> Vec<String> {
        self.0.clone()
    }
}

pub fn emit_cycles_changed<R: Runtime>(app: &tauri::AppHandle<R>, ids: &CycleIdSet) {
    if ids.is_empty() {
        return;
    }
    let _ = app.emit(
        CYCLES_CHANGED,
        CycleIdsPayload {
            cycle_ids: ids.ids(),
        },
    );
}

pub fn emit_tasks_changed<R: Runtime>(app: &tauri::AppHandle<R>, ids: &CycleIdSet) {
    if ids.is_empty() {
        return;
    }
    let _ = app.emit(
        TASKS_CHANGED,
        CycleIdsPayload {
            cycle_ids: ids.ids(),
        },
    );
}

pub fn emit_proposals_changed<R: Runtime>(app: &tauri::AppHandle<R>, cycle_id: &str) {
    let _ = app.emit(
        PROPOSALS_CHANGED,
        CycleIdPayload {
            cycle_id: cycle_id.to_string(),
        },
    );
}

/// Agent conversation change notice (add-ai-planning-core task 9.3): the
/// payload identifies the conversation; business data stays in the database.
pub const AGENT_CONVERSATION_UPDATED: &str = "agent:conversation_updated";

#[derive(Debug, Clone, Serialize)]
pub struct AgentConversationPayload {
    pub conversation_id: String,
    pub revision: i64,
}

pub fn emit_agent_conversation_updated<R: Runtime>(
    app: &tauri::AppHandle<R>,
    conversation_id: &str,
    revision: i64,
) {
    let _ = app.emit(
        AGENT_CONVERSATION_UPDATED,
        AgentConversationPayload {
            conversation_id: conversation_id.to_string(),
            revision,
        },
    );
}
